Lib/fontTools/ttLib/tables/otBase.py

```diff
@@ -398,11 +425,89 @@ def _gatherTables(self, tables, extTables, done):
 
 		selfTables.append(self)
 
-	def getAllData(self):
-		"""Assemble all data, including all subtables."""
+	def _gatherGraphForHarfbuzz(self, tables, obj_list, done, objidx, virtual_edges):
+		real_links = []
+		virtual_links = []
+		item_idx = objidx
+
+		# Merge virtual_links from parent
+		for idx in virtual_edges:
+			virtual_links.append((0, 0, idx))
+
+		sortCoverageLast = False
+		coverage_idx = 0
+		if hasattr(self, "sortCoverageLast"):
+			# Find coverage table
+			for i, item in enumerate(self.items):
+				if getattr(item, 'name', None) == "Coverage":
+					sortCoverageLast = True
+					if id(item) not in done:
+						coverage_idx = item_idx = item._gatherGraphForHarfbuzz(tables, obj_list, done, item_idx, virtual_edges)
+					else:
+						coverage_idx = done[id(item)]
+					virtual_edges.append(coverage_idx)
+					break
+
+		child_idx = 0
+		offset_pos = 0
+		for i, item in enumerate(self.items):
+			if hasattr(item, "getData"):
+				pos = offset_pos
+			elif hasattr(item, "getCountData"):
+				offset_pos += item.size
+				continue
+			else:
+				offset_pos = offset_pos + len(item)
+				continue
+
+			if id(item) not in done:
+				child_idx = item_idx = item._gatherGraphForHarfbuzz(tables, obj_list, done, item_idx, virtual_edges)
+			else:
+				child_idx = done[id(item)]
+                        
+			real_edge = (pos, item.offsetSize, child_idx)
+			real_links.append(real_edge)
+			offset_pos += item.offsetSize
+
+		tables.append(self)
+		obj_list.append((real_links,virtual_links))
+		item_idx += 1
+		done[id(self)] = item_idx
+		if sortCoverageLast:
+			virtual_edges.pop()
+
+		return item_idx
+
+	def getAllDataUsingHarfbuzz(self):
+		"""Assemble graph data and call Harfbuzz repacker to pack the table."""
```

Comment:
for the casual onlooker, maybe it's good if you add a link to the uharfbuzz code where you describe the format of the inputs to hb.repack() method, e.g. https://github.com/harfbuzz/uharfbuzz/blob/c11c7ba/src/uharfbuzz/_harfbuzz.pyx#L1149

Even maybe to that public doc that describes how the hb repacker algorithm works.